internal/pkg/source/source.go

```diff
@@ -114,51 +115,49 @@ func sourcesFromClosures(fn *ssa.Function, conf *config.Config, taggedFields fie
 }
 
 // sourcesFromBlocks finds Source values created by instructions within a function's body.
-func sourcesFromBlocks(fn *ssa.Function, conf *config.Config, taggedFields fieldtags.ResultType) []*Source {
+func sourcesFromBlocks(fn *ssa.Function, conf *config.Config, taggedFields fieldtags.ResultType, propagators fieldpropagator.ResultType) []*Source {
 	var sources []*Source
 	for _, b := range fn.Blocks {
 		for _, instr := range b.Instrs {
-			// This type switch is used to catch instructions that could produce sources.
-			// All instructions that do not match one of the cases will hit the "default"
-			// and they will not be examined any further.
+			var isSource bool
+
+			// Identify those instructions which produce a source.
 			switch v := instr.(type) {
-			// drop anything that doesn't match one of the following cases
-			default:
-				continue
-
-			// local variable or value returned from a call
-			case *ssa.Alloc, *ssa.Call:
-				if isProducedBySanitizer(v.(ssa.Value), conf) {
-					continue
-				}
+
+			// Values produced by sanitizers
+			case *ssa.Alloc:
+				isSource = !isProducedBySanitizer(v, conf) && IsSourceType(conf, taggedFields, v.Type())
+
+			// Return values produced by sanitizers cannot be sources,
+			case *ssa.Call:
+				isSource = !isProducedBySanitizer(v, conf) &&
+					(propagators.IsFieldPropagator(v) || IsSourceType(conf, taggedFields, v.Type()))
 
 			// An Extract is used to obtain a value from an instruction that returns multiple values.
 			// If the extracted value is a Pointer to a Source, it won't have an Alloc, so we need to
 			// identify the Source from the Extract.
 			case *ssa.Extract:
 				t := v.Tuple.Type().(*types.Tuple).At(v.Index).Type()
 				if _, ok := t.(*types.Pointer); ok && IsSourceType(conf, taggedFields, t) {
-					sources = append(sources, New(v))
+					isSource = true
 				}
-				continue
 
-			// value received from chan
+			// Values received from <-chan operations can be sources based on type
 			case *ssa.UnOp:
-				// not a <-chan operation
-				if v.Op != token.ARROW {
-					continue
+				isSource = v.Op == token.ARROW && IsSourceType(conf, taggedFields, v.Type())
+
+			// Identify sources by type in field access, element access of a collection, or map and chan collection creation.
+			// Array and slice collections are created via Alloc.  TODO: Then why is there a ssa.MakeSlice?
```

Comment:
The original comment:
```golang
// chan or map value (arrays and slices are created using Allocs)
case *ssa.MakeMap, *ssa.MakeChan:	
```
seems incorrect, or at least incomplete.  There is a `ssa.MakeSlice` instruction.  An array will necessarily have a corresponding alloc for the underlying array, but is that guarantee present for a `MakeSlice`?

_Well_ outside the scope of this PR now, but I figured I'd bring it up.

Reply:
Good catch. Judging from the code in `ssa/builder.go`, it looks like `MakeSlice` is produced when the length or capacity in a call to `make` is a variable. This is a blind spot in our testing. I'm not a huge fan of the naked TODO, but if you intend to follow this PR up with a fix right away, I suppose it's fine. If you don't have the cycles, I can make the fix.

Reply:
> I'm not a huge fan of the naked TODO,
Me either.  Even annotated ones bother be a bit.  I personally used them as a `DO NOT SUBMIT; TALK ABOUT THIS FIRST` marker.

As for the actual issue, it's another out-of-scope thing.  I'll chase it up with an Issue.

Reply:
There seems to be an issue with your markdown here, maybe a missing line between the quote and your reply?

> I personally used them as a DO NOT SUBMIT

+1, good to know. I also do that.